django/thunderstore/social/providers.py

```diff
@@ -0,0 +1,112 @@
+from typing import Dict
+
+import requests
+from django.conf import settings
+
+
+class BaseOauthHelper:
+    token = None
+
+    def __init__(self, code: str, redirect_url: str) -> None:
+        self.code = code
+        self.redirect_url = redirect_url
+
+    def complete_login(self) -> None:
+        """
+        Fetch access token with the temporary code and configured secret.
+        """
+        data = {
+            "client_id": self.CLIENT_ID,
+            "client_secret": self.CLIENT_SECRET,
+            "code": self.code,
+            "grant_type": "authorization_code",  # Required by Discord
+            "redirect_uri": self.redirect_url,  # Required by Discord
+        }
+        headers = {"Accept": "application/json"}
+
+        response = requests.post(self.OAUTH_URL, data, headers=headers)
+        response_data = response.json()
+        self.token = response_data["access_token"]
+
+    def get_user_info(self) -> Dict:
+        raise NotImplementedError()
+
+    def _fetch_from_api(self, url: str) -> Dict:
+        """
+        Use access token to fetch info from provider's public API.
+        """
+        if self.token is None:
+            raise Exception("No token found. Did you call .complete_login()?")
+
+        headers = {"Authorization": f"{self.AUTH_HEADER_KEYWORD} {self.token}"}
+
+        response = requests.get(url, headers=headers)
+        return response.json()
+
+
+class DiscordOauthHelper(BaseOauthHelper):
+    """
+    Helper class for Discord's OAuth authentication flow.
+    """
+
+    AUTH_HEADER_KEYWORD = "Bearer"
+    CLIENT_ID = settings.SOCIAL_AUTH_DISCORD_KEY
+    CLIENT_SECRET = settings.SOCIAL_AUTH_DISCORD_SECRET
+    OAUTH_URL = "https://discord.com/api/v8/oauth2/token"
+
+    def get_user_info(self) -> Dict:
+        """
+        Fetch user info from Discord's API using the access token.
+
+        https://github.com/discord/discord-api-docs/blob/main/docs/resources/User.md
+        """
+        data = self._fetch_from_api("https://discord.com/api/v8/users/@me")
+
+        return {
+            "email": data["email"],
+            "name": "",
```

Comment:
Assuming this (empty name) is the same as what current social auth does, but can you confirm?

Reply:
It is.